### DISCUSSION TITLE
Advent of TypeScript

### DISCUSSION THREAD

**Discussion by ayrock-dev** [typehero/typehero#1078] [Ideas]
# Idea Doc: Advent of TypeScript
🌲⭐️

## Goals

Advent of Code gets significant participation, fostering hype across a variety of social platforms.

- The [Advent of Code](https://adventofcode.com/) website
- Participation across GitHub
- [r/AdventOfCode](https://www.reddit.com/r/AdventOfCode/)
- Twitter, YouTube, and more

TypeHero stands to gain visibility and interest by publishing content for Advent of Code.

## Background

What is Advent of Code? Learn more from Eric Wastl at https://adventofcode.com/2023/about

## Naming

**Advent of TypeScript** is used as title of the event (Tweet, hero banner, etc) and **Twelve Days of TypeScript** is subtitle/tagline copy when presented in a hero/feature banner. **Twelve Days of TypeScript** represents the name of this year's event.

Next year can re-evaluate copy as-needed.

### Content (#1119)

The Twelve Days of TypeScript can be built out with 12 existing challenges on TypeHero, ramping from Beginner to Hard difficulty.

### Create Track (#1117)

Implement Advent of TypeScript as a track. Be sure to use the year (2023) in the title for the track.

![image](https://github.com/typehero/typehero/assets/7817695/c9a85c58-6d9a-4bbf-ae01-c555ad081c1f)

### Holiday Track UI (#1118)

UI for an Advent of TypeScript track can have extra winter holiday branding.

![image](https://github.com/typehero/typehero/assets/7817695/4fc13598-e980-4ff3-879b-171ab84701ee)

This should be hard-coded to display by configuring the Advent of TypeScript slug in a new Edge Config entry.

:bulb: **Note**: Use `holdayFoo` rather than `adventOfTypeScriptFoo` for shorter variable names, and does not lock-in the name of the holiday event in the code. However, the public-facing slug should be the full event name.

```json
{
  "holidayTracks": ["advent-of-typescript-2023"],
}
```

```ts
const isHolidayVariant = holidayTracks.includes(track.slug);
```

### Holiday Challenge Card UI (#1121)

On the Explore page, challenges in the Advent of TypeScript track can be rendered with a new card variant. 

🚧 WIP: Design in-progress.

### Holiday Hero/Feature Banner (#1122)

Advent of TypeScript can be advertised at the top of the Explore page. This banner funnels and directs users who land on TypeHero towards Advent of TypeScript.

🚧 WIP: Design in-progress.

## Areas of Work: Nice-To-Haves

Advent of Code has some cool incentives to join in. Participants earn :star:s along the way, publish answers/solutions on public GitHub profiles, and reveal special graphics, animation, and visuals throughout their Advent of Code experience.

TypeHero could offer similar incentives and shareable clout for users who participate in Advent of TypeScript.

### Badge (#1120)

TypeHero already has the general idea of profile badges on the roadmap for future milestones. A badge might look like badges earned on GitHub for getting started on GitHub and participating in events or hackathons. For example, badges on GitHub look something like the following.

![image](https://github.com/typehero/typehero/assets/7817695/6fb9e12f-22a6-467c-b060-3088ffffd8d7)
Reference: https://github.com/bautistaaa

While a fully-featured badge system requires significant work, hard-coded badges can be added by checking the number of completed challenges in the Advent of TypeScript track. 

List of badges:
 *  "Advent of TypeScript 2023 Participant" badge.
 *  "Advent of TypeScript 2023 Completionist" badge.

```ts
const isHoliday2023Participant = /* has at least one challenge in holiday track completed */;
const isHoliday2023Completionist = /* has all challenges in holiday track completed */;

return (
  {isHoliday2023Participant && <Holiday2023ParticipantBadge />}
  {isHoliday2023Completionist && <Holiday2023CompletionistBadge />}
);
```

Render badges on user's profile in new "Badges" card.

![Screenshot 2023-11-08 at 4 22 46 PM](https://github.com/typehero/typehero/assets/7817695/f6f6f262-d43c-445e-8e0a-719aa7fe53c8)

### User-Specific Open Graph (OG) Share URL (todo: break into issue)

During Advent of TypeScript, we might support a "Share URL" to share progress - or completion - along the Advent of Code track.

A user-specific share URL may highlight number of completed Advent of TypeScript track challenges, or render an extra-special "Completed"/100% badge which users may share on Twitter, Discord, and other social media sites.

Recent examples of OG images might look like

* TypeHero's own challenge OG links
* NextConf's ticket OG links

**Comment by bautistaaa** [comment]
I think we can pull this off pretty easy leveraging what we currently have. The issue remains in terms of content but I have no issue using challenges we currently have.

**Comment by ayrock-dev** [comment]
👍 sounds good, lemme pivot and revise the discussion above

changes:
* drop section on custom challenges/writing
* add section on feature/hero banner
* add question: how do we wanna tag this track to apply special styles/FX/visuals to it in the UI?

**Comment by maral** [comment]
Hi there, we do Advent of Typescript with my colleagues, thanks for a nice challenge (also great the tasks are finally getting harder :)). We were wondering if you were thinking of doing some kind of gamification like AoC has, i.e.:
- a leaderboard with the fastest times of each day
- private leaderboards for friends/teams to compete against each other
- some kind of points system like AoC has (although the public one is not very inclusive and doesn't let many people earn points :))

**Reply by bautistaaa** [reply]
that’s the plan. we pull this together in two weeks so we couldn’t get tooo crazy

**Reply by maral** [reply]
Perfect, thanks! I read through the plans / issues and I just didn't find any mention, so I wasn't sure.
